storage/manager.go

```diff
@@ -12,7 +12,7 @@ import (
 
 type Manager interface {
 	Get(ctx context.Context, collection string, key string, value interface{}) error
-	List(ctx context.Context, collection string, value interface{}, limit, offset int) error
```

Comment:
This `List` method is a generic. It can fetch policies, roles, and so on. Adding `member` here makes this interface function no longer generic but specific to the "list roles with members" method. Therefore, this change must be reversed.

Instead, what you could do, would be adding a `filter` property with type

```go
type Filters map[string]interface{}
```

and allowing filters to be passed here:

```go
List(ctx context.Context, collection string, value interface{}, filters Filters, limit, offset int) error
```

Then, in e.g. SQL you could construct a where clause using the filter, using the map key as the field name and the map value as the value for the field and making an "AND" clause if there are multiple filters.

Alternatively, you could even type the value part of the Filter, e.g.:

```go
type Filter struct {
  Value interface{}
  Foo string // whatever
}

type Filters map[string]Filter
```

Thank you!

Reply:
Sure, that makes sense. Although, I do have some reservations on why are CRUD functions generic here. We lose type assertions by passing the value as `interface{}`. Implementing a filter which matches against JSON document can lead to unintended consequences during run time.

Reply:
Yeah, you're right, I also overlooked how the query is constructed:

```
 		query = "SELECT document FROM rego_data WHERE collection=? AND (length(?)=0 OR json_extract(document,'$') LIKE ?) ORDER BY id ASC LIMIT ? OFFSET ?"
 	case dbal.DriverPostgreSQL:
 		query = "SELECT document FROM rego_data WHERE collection=? AND (length(?)=0 OR document::text LIKE ?) ORDER BY id ASC LIMIT ? OFFSET ?"
```

Since the database doesn't use any index here it will be as fast as loading these to memory and filtering them in Go. If we get performance issues there's always caching. Long-term we will replace this storage adapter anyways!

So how about doing this instead:

```go
type ListRequest struct {
	Collection string
	Value      interface{}
+.     FilterFunc(*ListRequest)
}

+ func (*l ListRequest) ListRequest Filter() *ListRequest {
+    if l.FilterFunc != nil {
+       l.FilterFunc(l)
+   }
+   return l
+ }


func (h *Handler) List(factory func(context.Context, *http.Request, httprouter.Params) (*ListRequest, error)) httprouter.Handle {
	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
		ctx := r.Context()
		l, err := factory(ctx, r, ps)
		if err != nil {
			h.h.WriteError(w, r, err)
			return
		}

		limit, offset := pagination.Parse(r, 100, 0, 500)
		if err := h.s.List(ctx, l.Collection, l.Value, limit, offset); err != nil {
			h.h.WriteError(w, r, err)
			return
		}

-		h.h.Write(w, r, l.Value)
+		h.h.Write(w, r, l.Filter().Value)
	}
}
```

And use the Filter function to sort for roles by members? We would need to do a type assertion here but it's safe to say that the type will always validate to true, and if not it's a programming error (should thus panic).

Reply:
I like this approach more than the previous. I also like the approach of just indexing the JSON data using the approach mentioned [here](https://docs.microsoft.com/en-us/sql/relational-databases/json/index-json-data?view=sql-server-2017). Regarding the type-assertion; the problem is not type-assertion after `List()` returns, it is before `Upsert()` is called as I can pass `value interface{} `as literally anything, but in reality, this `value interface{} `should point to `type Role`. In the interest of time, I can move forward with the filter approach.